types/wlr_relative_pointer_v1.c

```diff
@@ -0,0 +1,192 @@
+#include <stdlib.h>
+#include <assert.h>
+#include <inttypes.h>
+#include <wlr/util/log.h>
+#include <util/signal.h>
+#include <wlr/types/wlr_relative_pointer_v1.h>
+#include <wlr/types/wlr_seat.h>
+#include "wayland-util.h"
+#include "wayland-server.h"
+#include "relative-pointer-unstable-v1-protocol.h"
+
+static const struct zwp_relative_pointer_manager_v1_interface relative_pointer_manager_v1_impl;
+static const struct zwp_relative_pointer_v1_interface relative_pointer_v1_impl;
+
+
+/**
+ * Callback functions
+ */
+
+static void relative_pointer_manager_v1_handle_resource_destroy(struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+
+static void relative_pointer_v1_handle_resource_destroy(struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+
+	struct wlr_relative_pointer_v1 *relative_pointer = wl_resource_get_user_data(resource);
+	free(relative_pointer);
+}
+
+
+static void relative_pointer_manager_v1_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 manager unbound from client %p",
+		client);
+}
+
+
+static void relative_pointer_manager_v1_handle_get_relative_pointer(struct wl_client *client,
+		struct wl_resource *resource, uint32_t id, struct wl_resource *pointer) {
+	struct wlr_seat_client *seat_client =
+		wlr_seat_client_from_pointer_resource(pointer);
+	assert(seat_client->client == client);
+
+	struct wlr_relative_pointer_v1 *relative_pointer =
+		calloc(1, sizeof(struct wlr_relative_pointer_v1));
+	if (relative_pointer == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wl_resource *relative_pointer_resource = wl_resource_create(client,
+		&zwp_relative_pointer_v1_interface, wl_resource_get_version(resource), id);
+	if (relative_pointer_resource == NULL) {
+		free(relative_pointer);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	relative_pointer->client = client;
+	relative_pointer->resource = relative_pointer_resource;
+	relative_pointer->pointer = (struct wl_pointer *) pointer;
+
+	wl_signal_init(&relative_pointer->events.destroy);
+
+	wl_list_insert(&seat_client->relative_pointers,
+		wl_resource_get_link(relative_pointer_resource));
+
+	wl_resource_set_implementation(relative_pointer_resource, &relative_pointer_v1_impl,
+		relative_pointer, relative_pointer_v1_handle_resource_destroy);
+
+	struct wlr_relative_pointer_manager_v1 *relative_pointer_manager =
+		wl_resource_get_user_data(resource);
```

Comment:
There should be a safe helper instead of directly using `wl_resource_get_user_data` (this one can be local to this file)